VAC/RAW/ ETH-MLS-OFFCHAIN RFC multi-steward support (#193)
vacp2p/rfc-index: vac/raw/eth-mls-offchain.md

This is restricted by [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) as if there is pending proposal,
the application message should be cut.
Note that: Since the MLS is based on servers, this delay between proposal and commit messages are very small.

`Commit message:` After members receive the proposals regarding group changes,
the committer, who may be any member of the group, as specified in  [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/),
generates the necessary key material for the next epoch, including the appropriate welcome messages
for new joiners and new entropy for removed members. In this RFC, the committers only MUST be stewards.

### de-MLS Objects

This section presents the de-MLS objects:

`Voting proposal`: Similar to MLS proposals, but processed only if approved through a voting process.
They function as application messages in the MLS group,
allowing the steward to collect them without halting the protocol.
There are three types of `voting proposal` according to the type of consensus as in shown Consensus Types section,

> these are, `commit proposal`, `steward election proposal` and `emercency criteria proposal`.

Reviewer: ```suggestion
these are, `commit proposal`, `steward election proposal` and `emergency criteria proposal`.
```

`Epoch steward`: The steward assigned to commit in `epoch E` according to the steward list.
Holds the primary responsibility for creating commit in that epoch.

`Backup steward`: The steward next in line after the `epoch steward` on the `steward list` in `epoch E`.
Only becomes active if the `epoch steward` is malicious or fails,
in which case it completes the commitment phase.
If unused in `epoch E`, it automatically becomes the `epoch steward` in `epoch E+1`.

`Steward list`: It is an ordered list that consists of authorized stewards' `member id`s which belongs

> to stewards who eligible to create commits when a particular steward commit order should be defined beforehand.

Author: Indeed, I clarified and gave a mini example [here](https://github.com/vacp2p/rfc-index/pull/193/commits/35e1610966acf2469bc836c40d7f7c2798df5749) and [here](https://github.com/vacp2p/rfc-index/pull/193/commits/de5241cccc657c88c1d4dbf3b43a3be85abf519d)

Decentralization has already been achieved in the previous section.
However, to improve availability and ensure censorship resistance,
the single steward protocol is extended to a multi steward architecture.
In this design, each epoch is coordinated by a designated steward,
operating under the same protocol as the single steward model.
Thus, the multi steward approach primarily defines how steward roles
rotate across epochs while preserving the underlying structure and logic of the original protocol.
Two variants of the multi steward design are introduced to address different system requirements.

### Consensus Types

Consensus is agnostic with its payload; therefore, it can be used for various purposes.
Note that each message for the consensus of proposals is an `application message` in the MLS object section.
It is used in three ways as follows:

> 1. `Commit proposal`: Each commit message MUST be committed with its YES vote by a specific steward.

Author: You know, in single steward we had only one steward and updates all commit messages when the proposal is voted YES right. In here, it is the same, only difference is steward must be epoch steward. Otherwise, if we have many stewards that want to commit, I am considering that might be conflicts. 
[Here is a clarification](https://github.com/vacp2p/rfc-index/pull/193/commits/23d31f128a11dd060aedcfcb352da8d99c990568).

and SHOULD be initialized by any member during the corresponding epoch.
The Proposal.payload field MUST represent the ordered identities of the proposed stewards.
Each steward election proposal MUST be verified and finalized through the consensus process
so that members can identify which steward will be responsible in each epoch
and detect any unauthorized steward commits.
3. `Emergency criteria proposal`: If there is a malicious member or steward,
this event MUST be voted on to finalize it.
If this returns YES, the next epoch MUST include the removal of the member or steward.
Proposal.payload MUST consist of the evidence of the dishonesty as described in the Steward violation list,
and the identifier of the malicious member or steward.
This proposal can be created by any member in any epoch.

The order of consensus proposal messages is important to achieving a consistent result.
Therefore, messages MUST be prioritized by type in the following order, from highest to lowest priority:

- `Emergency criteria proposal`

- `Steward election proposal`

- `Commit proposal`

This means that if a higher-priority consensus proposal is present in the network,

> lower-priority messages MUST be withheld from transmission until the higher-priority proposals have been finalized.

Author: Thank you for very detail proposals. let's have another round then consensus on the mechanism then I can add them as another section. 

> All should stewards collect incoming proposal

It is a bit hard not only the decryption, also the all stewards needs to be online all the time. Also we might assume some stewards can be offline or network break since our assumption is this works on unreliable network. But, we can make it, few stewards can get keyPackages `epoch `and `backup stewards`. Since, the non-member nodes can see the public key of stewards in the current list in welcome topic, then encrypt then send the keyPackages these two steward.

> If we get emergency proposal, we move group to Waiting State 

I quite like this idea, we can let all know about the waiting state with its reason, since the reason is transparent all the group I think this action is verifiable. 

> Q: who should send emergency proposal in case that steward for example just have a bad connection and can't create commit right now. Do we have any solution in waku for such ping?

Waku has SDS but not sure %100 but Waku has ping since it is build on top of nim libp2p and nim libp2p has default action as mostly for other p2p protocols. I think, we do not need to restrict the sender or emergency proposal everyone can init it, we can maybe say that the application can restrict instead of emergency proposal DoS. 

> if we previously have some proposals from this steward , do we reject it or no?

I think, previous proposals can stay since at that time this steward was honest, but it is not now.

> after everyone get it we change state from emergency to waitnig and start usual epoch, so that means we handle all previously collected proposals. And here we perhaps can skip time gap for emergency if we take to much

This is very cool, it means we just transfer the proposals to next epoch so we do not need to reject them.

> steward not create commit when new epoch starts because he is offline - we need additional timeout in epoch to understand that steward not response and make next steward as main. Maybe even not to kick this user from the group

It is easy to init a emergency proposal when there is concrete violation but, for inactivity, there is a [comment ](https://github.com/vacp2p/rfc-index/pull/193#discussion_r2449137250) about the time-epoch issue. I think we have no exact time interval for epoch right such as in seconds. The epoch is changed when the steward release a commit message. So, we need to have some time interval to define the inactivity situtation somehow, otherwise how we can understand that the steward is silent (malicious or network issue does not matter).

### Steward list creation

The `steward list` size as `sn` is determined when the group is created.
The index of the slots shows that the Epoch and `member id`s are stored.
The next in line steward for the `epoch E` is named as `epoch steward`, which has index E.

> And the subsequent steward in the `epoch E` is named as the `backup steward`.

Reviewer: little bit hard to understand, maybe example will help

but do not take part in commit decision by `commit proposal`.
This structure enables faster coordination since consensus is achieved within a smaller group of stewards.
It is particularly suitable for large user groups, where involving every member in each decision would be impractical.

The flow is similar to the big consensus including the `steward list` finalization with all members consensus
only the difference here, the commit messages requires `commit proposal` only among the stewards.

## Steward violation list

A steward’s activity is called a violation if the action is one or more of the following:

1. Broken commit: The steward releases a different commit message from the voted `commit proposal`.
This activity is identified by the `members` since the [MLS RFC 9420](https://datatracker.ietf.org/doc/rfc9420/) provides the methods
that members can use to identify the broken commit messages that are possible in a few situations,
such as commit and proposal incompatibility. Specifically, the broken commit can arise as follows:
The commit belongs to the earlier epoch. The commit message should equal the latest epoch,

> or the commit needs to be compatible with the previous epoch’s `MLS proposal`.

Reviewer: ```suggestion
    1. The commit belongs to the earlier epoch. 
    2. The commit message should equal the latest epoch
    3. The commit needs to be compatible with the previous epoch’s `MLS proposal`.
```

the single steward protocol is extended to a multi steward architecture.
In this design, each epoch is coordinated by a designated steward,
operating under the same protocol as the single steward model.
Thus, the multi steward approach primarily defines how steward roles
rotate across epochs while preserving the underlying structure and logic of the original protocol.
Two variants of the multi steward design are introduced to address different system requirements.

### Consensus Types

Consensus is agnostic with its payload; therefore, it can be used for various purposes.
Note that each message for the consensus of proposals is an `application message` in the MLS object section.
It is used in three ways as follows:

1. `Commit proposal`:  It is the proposal instance that is specified in Creating Voting Proposal section
with `Proposal.payload` MUST show the commit request from `members`.
Any member MAY create this proposal in any epoch and `epoch steward` MUST collect and commit YES voted proposals.
This is the only proposal type common to both single steward and multi steward designs.
2. `Steward election proposal`: This is the process that finalizes the `steward list`,

> which sets and orders stewards responsible for creating commits over a predefined number of epochs.

Reviewer 2: If a steward is removed due to an emergency proposal, what happens to the steward list? Does a new steward election happen?

This proposal is created periodically, once every number of epochs equal to the steward list size,
and SHOULD be initialized by any member during the corresponding epoch.
The Proposal.payload field MUST represent the ordered identities of the proposed stewards.
Each steward election proposal MUST be verified and finalized through the consensus process
so that members can identify which steward will be responsible in each epoch
and detect any unauthorized steward commits.
3. `Emergency criteria proposal`: If there is a malicious member or steward,
this event MUST be voted on to finalize it.

> If this returns YES, the next epoch MUST include the removal of the member or steward.

Reviewer 2: >the next epoch MUST include the removal of the member or steward.
Is this removal from the group or from the steward list?

If Steward List: Won't that create ambiguity over who the steward is supposed to be?
If Group: Inactivity is listed as reason 3 in the Steward violation list. If stewards are randomly assigned, this means that members could be removed from groups due to network connectivity issues which has usability and security implications

- `Steward election proposal`

- `Commit proposal`

This means that if a higher-priority consensus proposal is present in the network,
lower-priority messages MUST be withheld from transmission until the higher-priority proposals have been finalized.

### Steward list creation

The `steward list` size as `sn` is determined when the group is created.
The index of the slots shows epoch info and value of index shows `member id`s.
The next in line steward for the `epoch E` is named as `epoch steward`, which has index E.
And the subsequent steward in the `epoch E` is named as the `backup steward`.
For example, let's assume steward list is (S3, S2, S1) if in the previous epoch the roles were
(`backup steward`: S2, `epoch steward`: S1), then in the next epoch they become
(`backup steward`: S3, `epoch steward`: S2) by shifting.

If the `epoch steward` is honest, the `backup steward` does not involve the process in epoch,
and the `backup steward` will be the `epoch steward` within the `epoch E+1`.

If the `epoch steward` is malicious, the `backup steward` is involved in the commitment phase in `epoch E`
and the former steward becomes the `backup steward` in `epoch E`.

Liveness criteria:

> Once the active `steward list` has completed its assigned epoch,

Author: Yes, in fact it is a range starts, like a bijective map with epochs and stewards. In here, I wanted to say "practically ends with last epoch as a single one" but this works also.

members MUST proceed to elect the next set of stewards
(which MAY include some or all of the previous members).
This election is conducted through a type 2 consensus procedure, `steward election proposal`.

A `Steward election proposal` is considered valid only if the resulting `steward list`
is produced through a deterministic process that ensures an unbiased distribution of steward assignments,
since allowing bias could enable a malicious participant to manipulate the list
and retain control within a favored group for multiple epochs.

The list MUST be composed of the first `sn` members from the member list,

> sorted according to the ascending value of `SHA256(epoch E || member id)`,

Reviewer 2: Not critical, but worth noting. 
Using only static data to derive the list means that malicious actors could synthesis a set of member_ids which provide many epochs of stewardship in the future. I don't think incorporating ratcheted data is worth the complexity at this point; but worth noting in the security model.

where `epoch E` is the epoch in which the election proposal is initiated.

> Any proposal with a list that does not adhere to this generation method MUST be rejected by all members.

Reviewer 2: I love how this is deterministic as it means every client can derive the same steward list. However I'm curious about the case where the current steward list contains only "Stale" members. (Members who are are valid members of the group, but remain offline for an extended period of time; Either due to unavailability, or a fault causing a desynchronization in state)
- The larger the group the more likely that stale clients exist
- During this time proposals build up, but there is no steward available to generate valid commits.
- Lack of commits results in a partial deadlock, until the era expires and a new steward election proposal occurs and an "online" steward is selected.

Is the intention that applications must prune inactive members to avoid this case?

(`backup steward`: S2, `epoch steward`: S1), then in the next epoch they become
(`backup steward`: S3, `epoch steward`: S2) by shifting.

If the `epoch steward` is honest, the `backup steward` does not involve the process in epoch,
and the `backup steward` will be the `epoch steward` within the `epoch E+1`.

If the `epoch steward` is malicious, the `backup steward` is involved in the commitment phase in `epoch E`
and the former steward becomes the `backup steward` in `epoch E`.

Liveness criteria:

Once the active `steward list` has completed its assigned epochs,

members MUST proceed to elect the next set of stewards
(which MAY include some or all of the previous members).
This election is conducted through a type 2 consensus procedure, `steward election proposal`.

A `Steward election proposal` is considered valid only if the resulting `steward list`
is produced through a deterministic process that ensures an unbiased distribution of steward assignments,
since allowing bias could enable a malicious participant to manipulate the list
and retain control within a favored group for multiple epochs.

The list MUST be composed of the first `sn_min` members including previous stewards,

> sorted according to the ascending value of `SHA256(epoch E || member id || group id)`,

Reviewer 2: I think this section could use some added clarity, to make it super clear to readers.

Specifically that it is made by computing a hash **for each member of the group**, and sorting the results into the steward list.

where `epoch E` is the epoch in which the election proposal is initiated,
and `group id` for shuffling the list across the different groups.
Any proposal with a list that does not adhere to this generation method MUST be rejected by all members.

> We assume that there are no recurring entries in `SHA256(epoch E || member id)`, since the SHA256 outputs are unique

Reviewer 2: update to match above

```suggestion
We assume that there are no recurring entries in `SHA256(epoch E || member id || group id)`, since the SHA256 outputs are unique
```
